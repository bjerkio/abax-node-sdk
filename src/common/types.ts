export interface QueryEnvelope<C> {
  query: C;
}

export interface BodyEnvelope<C> {
  body: C;
}
