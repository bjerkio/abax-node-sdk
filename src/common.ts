import { addMinutes, format, set } from 'date-fns';
import type { z } from 'zod';

export function startOfTheNextMinute(fromDate?: Date): Date {
  const now = fromDate ?? new Date();

  const zeroSeconds = set(now, {
    seconds: 0,
    milliseconds: 0,
  });

  return addMinutes(zeroSeconds, 1);
}

/**
 *
 * @param data search parameters
 * @returns the parameters formatted as URLSearchParams
 *
 * @example  makeSearchParams({ foo: 'bar', baz: 42 }) => 'foo=bar&baz=42'
 *
 * Create search parameters from a flat object.
 */
/** Creates search params from a one-dimensional object. Supports array values. */
export function makeSearchParams(
  data: Record<string, unknown>,
): URLSearchParams {
  const params = new URLSearchParams();
  Object.entries(data).forEach(([key, value]) => {
    if (value) {
      if (Object.prototype.toString.call(value) === '[object Array]') {
        Object.values(value).forEach(v => {
          params.append(key, makeStringFromSeachParam(v));
        });
      } else {
        params.append(key, makeStringFromSeachParam(value));
      }
    }
  });

  return params;
}

function makeStringFromSeachParam(value: unknown): string {
  if (value instanceof Date) {
    return format(value, 'yyyy-MM-dd');
  }

  return String(value);
}

export function withZod<T extends z.ZodTypeAny, Output = z.infer<T>>(
  schema: T,
): (data: unknown) => Output {
  return data => {
    return schema.parse(data) as Output;
  };
}
