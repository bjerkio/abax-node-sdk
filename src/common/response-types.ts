import { z } from 'zod';
import { camelize } from './types';

export function listResponseEnvelope<TMessageType extends z.ZodType>(
  zodType: TMessageType,
) {
  return z
    .object({
      page: z.number(),
      pageSize: z.number(),
      items: z.array(zodType),
    })
    .transform(camelize);
}
