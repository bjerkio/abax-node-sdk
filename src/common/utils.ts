import { addMinutes, set } from 'date-fns';
import type { z } from 'zod';

export function withZod<T extends z.ZodTypeAny, Output = z.infer<T>>(
  schema: T,
): (data: unknown) => Output {
  return data => {
    return schema.parse(data) as Output;
  };
}

/** Make URL params in query  */
export function makeQuery(data: {
  query: Record<string, unknown> | undefined;
}): URLSearchParams {
  const params = new URLSearchParams();

  if (!data.query) {
    return params;
  }

  Object.entries(data.query).forEach(([key, value]) => {
    if (value) {
      params.append(key, String(value));
    }
  });

  return params;
}

/** Make URL params in body */
export function makeBody(data: {
  body: Record<string, unknown>;
}): URLSearchParams {
  const params = new URLSearchParams();

  Object.entries(data.body).forEach(([key, value]) => {
    if (value) {
      params.append(key, String(value));
    }
  });

  return params;
}

export function startOfTheNextMinute(fromDate?: Date): Date {
  const now = fromDate ?? new Date();

  const zeroSeconds = set(now, {
    seconds: 0,
    milliseconds: 0,
  });

  return addMinutes(zeroSeconds, 1);
}
