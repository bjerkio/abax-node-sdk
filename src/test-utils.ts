import { MockAgent, type Interceptable, setGlobalDispatcher } from 'undici';
import { AbaxClient } from './abax-client';

export function initialiseClientAndMockPool(): {
  client: AbaxClient;
  mockPool: Interceptable;
} {
  const mockAgent = new MockAgent();
  mockAgent.disableNetConnect();
  setGlobalDispatcher(mockAgent);

  const mockPool = mockAgent.get('https://api.abax.cloud');

  const client = new AbaxClient({
    baseUrl: 'production',
    apiKey: 'kanonball!',
  });

  return { client, mockPool };
}
