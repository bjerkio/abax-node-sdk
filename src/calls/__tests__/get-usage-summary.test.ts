import { describe, expect, it } from 'vitest';
import { initialiseClientAndMockPool } from '../../test-utils';
import { makeSearchParams } from '../../common';

describe('get-usage-summary', () => {
  it('should return equipment', async () => {
    const { client, mockPool } = initialiseClientAndMockPool();

    const searchParams = makeSearchParams({
      from: new Date('2024-01-01'),
      to: new Date('2024-02-28'),
    })
    const path = `/v1/vehicles/394f39098dee561fb248b443e327c790/usage-summary?${searchParams.toString()}`;
    mockPool
      .intercept({
        path: path,
        method: 'GET',
      })
      .reply(200, {
        private_usage_summary: {
          distance_driven_in_meters: 1500,
          total_toll_stations_passed: 2,
          total_road_toll_cost: 40,
        },
        corporate_usage_summary: {
          distance_driven_in_meters: 468,
          total_toll_stations_passed: 4,
          total_road_toll_cost: 80,
        },
      })
      .times(1);
    
    const usageSummary = await client.listUsageSummary({
      vehicleId: '394f39098dee561fb248b443e327c790',
      dateFrom: new Date('2024-01-01'),
      dateTo: new Date('2024-02-28'),
    });

    expect(usageSummary).toMatchSnapshot();
  });
});
