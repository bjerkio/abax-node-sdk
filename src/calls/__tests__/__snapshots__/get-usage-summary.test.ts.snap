// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get-usage-summary > should return equipment 1`] = `
{
  "corporateUsageSummary": {
    "distanceDrivenInMeters": 468,
    "totalRoadTollCost": 80,
    "totalTollStationsPassed": 4,
  },
  "privateUsageSummary": {
    "distanceDrivenInMeters": 1500,
    "totalRoadTollCost": 40,
    "totalTollStationsPassed": 2,
  },
}
`;
